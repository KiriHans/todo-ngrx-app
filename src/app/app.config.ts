import { ApplicationConfig, isDevMode } from '@angular/core';
import { provideRouter } from '@angular/router';

import { provideAnimations } from '@angular/platform-browser/animations';

import { routes } from './app.routes';
import { provideStore } from '@ngrx/store';
import { provideStoreDevtools } from '@ngrx/store-devtools';
import { appReducers, metaReducers } from './app.reducer';

export const appConfig: ApplicationConfig = {
  providers: [
    provideRouter(routes),
    provideAnimations(),
    provideStore(appReducers, { metaReducers }),
    provideStoreDevtools({ maxAge: 25, logOnly: !isDevMode() }),
  ],
};
