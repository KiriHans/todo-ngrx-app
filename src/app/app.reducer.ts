import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import { TodoItem } from './todo-page/models/todo-item.model';
import { todoReducer } from './todo-page/store/reducers/todo.reducer';
import { filterReducer } from './todo-page/store/reducers/filter.reducer';
import { ValidFilters } from './todo-page/types/valid-filter.type';
import { storageMetaReducer } from './todo-page/store/reducers/todo.metareducer';

export interface AppState {
  todos: TodoItem[];
  filter: ValidFilters;
}

export const appReducers: ActionReducerMap<AppState> = {
  todos: todoReducer,
  filter: filterReducer,
};

export const metaReducers: MetaReducer[] = [storageMetaReducer];
