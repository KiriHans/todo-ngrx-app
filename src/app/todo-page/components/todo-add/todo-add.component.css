:host{
  --elevated-card-color: rgb(236, 236, 236);
  --elevation-card: 0px 2px 3px -1px rgba(0, 0, 0, .2), 0px 1px 2px 0px rgba(0, 0, 0, .14), 0px 1px 3px 0px rgba(0, 0, 0, .12);
  --border-top-radius: 4px;

}

:host ::ng-deep :is(.field-add-todo) .mat-mdc-form-field-subscript-wrapper {
  display: none;
}

.field-add-todo{
  width: 100%;
  background-color: #FFF;
  border-color: var(--elevated-card-color);
  box-shadow: var(--elevation-card);

  border-top-left-radius: var(--border-top-radius);
  border-top-right-radius: var(--border-top-radius);
}

span{
  position: relative;
  display: flex;
  align-items: center;
  padding-right: 1rem;
}

span > mat-icon{
  position: relative;
  top: 0.5rem;
}

.completed{
  color: #ff4081;
  
}

button{
  margin: 0 0.5rem;
}

.add-button{
  margin-right: 0.7rem;

}

.toggle-icon{
  transform: scale(1.7);
}

.add-icon{
  transform: scale(1.3);

}


.material-symbols-outlined {
  font-variation-settings:
  'FILL' 0,
  'wght' 400,
  'GRAD' 0,
  'opsz' 48
}
