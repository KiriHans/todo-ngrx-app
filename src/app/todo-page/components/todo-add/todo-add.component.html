<mat-form-field class="field-add-todo">
  <button (click)="toggleAllTodos($event)" matIconPrefix mat-icon-button>
    <mat-icon *ngIf="toggleAll; else completed" class="toggle-icon completed"> toggle_on </mat-icon>
    <ng-template #completed>
      <mat-icon class="toggle-icon no-completed">
        <span class="material-symbols-outlined"> toggle_off </span>
      </mat-icon>
    </ng-template>
  </button>

  <input
    matInput
    class="input-create"
    placeholder="What do you want to do?"
    [formControl]="todoAdd"
    (keyup.enter)="add()"
  />

  <button class="add-button" (click)="add()" matIconSuffix mat-icon-button>
    <mat-icon class="material-symbols-outlined add-icon"> add_outline </mat-icon>
  </button>
</mat-form-field>
