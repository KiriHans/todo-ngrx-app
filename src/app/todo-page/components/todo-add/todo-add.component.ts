import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatInputModule } from '@angular/material/input';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';
import { FormControl, ReactiveFormsModule, Validators } from '@angular/forms';
import { Store } from '@ngrx/store';
import { MatCheckboxModule } from '@angular/material/checkbox';

import * as TodoActions from '../../store/actions/todo.actions';

@Component({
  selector: 'app-todo-add',
  standalone: true,
  imports: [
    CommonModule,
    MatInputModule,
    MatIconModule,
    MatButtonModule,
    ReactiveFormsModule,
    MatCheckboxModule,
  ],
  templateUrl: './todo-add.component.html',
  styleUrls: ['./todo-add.component.css'],
})
export class TodoAddComponent {
  todoAdd: FormControl;
  toggleAll = false;

  constructor(private store: Store) {
    this.todoAdd = new FormControl('', Validators.required);
  }

  add(): void {
    if (this.todoAdd.invalid) return;
    this.store.dispatch(TodoActions.createTodo({ text: this.todoAdd.value }));
    this.todoAdd.reset();
  }

  toggleAllTodos(event: Event): void {
    event.stopPropagation();
    this.toggleAll = !this.toggleAll;
    this.store.dispatch(TodoActions.toggleAllTodos({ completed: this.toggleAll }));
  }
}
