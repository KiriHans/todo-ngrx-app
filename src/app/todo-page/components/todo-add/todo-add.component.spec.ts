import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
import { HarnessLoader, TestKey } from '@angular/cdk/testing';

import { MatButtonHarness } from '@angular/material/button/testing';
import { MatInputHarness } from '@angular/material/input/testing';

import { TodoAddComponent } from './todo-add.component';

import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { AppState } from 'src/app/app.reducer';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { DebugElement } from '@angular/core';

import { By } from '@angular/platform-browser';
import { Action } from '@ngrx/store';
import { CommonModule } from '@angular/common';

describe('TodoAddComponent', () => {
  let component: TodoAddComponent;
  let fixture: ComponentFixture<TodoAddComponent>;
  let el: DebugElement;
  let compiled: HTMLElement;

  let store: MockStore<AppState>;

  let harnessLoader: HarnessLoader;

  let dispatchSpy: jest.SpyInstance<void, [action: Action], unknown>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [CommonModule, TodoAddComponent, NoopAnimationsModule, ReactiveFormsModule],
      providers: [provideMockStore()],
    });
    fixture = TestBed.createComponent(TodoAddComponent);

    component = fixture.componentInstance;
    el = fixture.debugElement;
    compiled = fixture.nativeElement;

    store = TestBed.inject(MockStore);

    fixture.detectChanges();
  });

  beforeEach(() => {
    dispatchSpy = jest.spyOn(store, 'dispatch');

    harnessLoader = TestbedHarnessEnvironment.loader(fixture);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should render a form field with the class "field-add-todo"', () => {
    const fieldAdd = el.query(By.css('.field-add-todo'));

    expect(fieldAdd).toBeTruthy;
  });

  it("should render an input with it's respective prefix and suffix buttons", () => {
    const toggleAllButton = el.query(By.css('button[matIconPrefix]'));
    const toggleAllButtonText = (toggleAllButton.nativeElement as HTMLButtonElement).textContent;

    const inputTodo = el.query(By.css('input.input-create'));
    const inputPlaceholder = (inputTodo.nativeElement as HTMLInputElement).placeholder;

    const AddButton = el.query(By.css('button[matIconSuffix]'));
    const AddButtonText = (AddButton.nativeElement as HTMLButtonElement).textContent;

    expect(toggleAllButton).toBeTruthy();
    expect(toggleAllButtonText).toContain('toggle_off');

    expect(inputTodo).toBeTruthy();
    expect(inputPlaceholder).toBe('What do you want to do?');

    expect(AddButton).toBeTruthy();
    expect(AddButtonText).toContain('add_outline');
  });

  it('should initialize respective form control', () => {
    expect(component.todoAdd).toBeInstanceOf(FormControl);
  });

  it('should toggle all todo items when toggle all button is pressed', async () => {
    const toggleAllButton = await harnessLoader.getHarness(
      MatButtonHarness.with({ selector: 'button[matIconPrefix]' })
    );

    expect(component.toggleAll).toBeFalsy();

    await toggleAllButton.click();

    expect(component.toggleAll).toBeTruthy();
  });

  it('should dispatch the respective action when pressed enter while input is in focus', async () => {
    const addSpy = jest.spyOn(component, 'add');

    component.todoAdd.setValue('test');

    const inputTodo = await harnessLoader.getHarness(
      MatInputHarness.with({ selector: 'input.input-create' })
    );

    await (await inputTodo.host()).sendKeys(TestKey.ENTER);

    expect(addSpy).toHaveBeenCalled();
    expect(dispatchSpy).toHaveBeenCalled();
    expect(component.todoAdd.value).toBeNull();
  });

  it('should not dispatch the respective action when input is invalid', async () => {
    const addSpy = jest.spyOn(component, 'add');

    const inputTodo = await harnessLoader.getHarness(
      MatInputHarness.with({ selector: 'input.input-create' })
    );

    await (await inputTodo.host()).sendKeys(TestKey.ENTER);

    expect(addSpy).toHaveBeenCalled();
    expect(dispatchSpy).not.toHaveBeenCalled();
  });

  it('should dispatch the respective action when add button is pressed', async () => {
    const addSpy = jest.spyOn(component, 'add');

    component.todoAdd.setValue('test');

    const addTodoButton = await harnessLoader.getHarness(
      MatButtonHarness.with({ selector: 'button[matIconSuffix]' })
    );

    await addTodoButton.click();

    expect(addSpy).toHaveBeenCalled();
    expect(dispatchSpy).toHaveBeenCalled();
  });

  it('should change add button image when pressed', async () => {
    let toggleAllIcon = el.query(By.css('.toggle-icon'));
    const toggleButton = el.query(By.css('button[matIconPrefix]'));

    expect(toggleAllIcon.nativeElement.textContent.trim()).toBe('toggle_off');

    toggleButton.nativeElement.click();

    fixture.detectChanges();

    toggleAllIcon = el.query(By.css('.toggle-icon'));

    expect(toggleAllIcon.nativeElement.textContent.trim()).toBe('toggle_on');
  });
});
