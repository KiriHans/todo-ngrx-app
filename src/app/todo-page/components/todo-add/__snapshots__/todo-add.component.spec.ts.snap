// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAddComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <mat-form-field
    class="mat-mdc-form-field field-add-todo mat-mdc-form-field-type-mat-input mat-mdc-form-field-has-icon-prefix mat-mdc-form-field-has-icon-suffix mat-form-field-no-animations mat-form-field-appearance-fill mat-primary ng-untouched ng-pristine ng-invalid"
  >
    
    <div
      class="mat-mdc-text-field-wrapper mdc-text-field mdc-text-field--filled mdc-text-field--no-label"
    >
      <div
        class="mat-mdc-form-field-focus-overlay"
      />
      
      <div
        class="mat-mdc-form-field-flex"
      >
        
        <div
          class="mat-mdc-form-field-icon-prefix"
        >
          <button
            class="mdc-icon-button mat-mdc-icon-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
            mat-button-ripple-uninitialized=""
            mat-icon-button=""
            maticonprefix=""
          >
            <span
              class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
            />
            <mat-icon
              aria-hidden="true"
              class="mat-icon notranslate toggle-icon no-completed material-icons mat-ligature-font mat-icon-no-color"
              data-mat-icon-type="font"
              role="img"
            >
              <span
                class="material-symbols-outlined"
              >
                 toggle_off 
              </span>
            </mat-icon>
            
            
            <span
              class="mat-mdc-focus-indicator"
            />
            <span
              class="mat-mdc-button-touch-target"
            />
          </button>
        </div>
        
        
        <div
          class="mat-mdc-form-field-infix"
        >
          
          
          
          <input
            aria-required="true"
            class="mat-mdc-input-element input-create mat-mdc-form-field-input-control mdc-text-field__input ng-untouched ng-pristine ng-invalid cdk-text-field-autofill-monitored"
            matinput=""
            placeholder="What do you want to do?"
            required=""
          />
        </div>
        
        <div
          class="mat-mdc-form-field-icon-suffix"
        >
          <button
            class="add-button mdc-icon-button mat-mdc-icon-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
            mat-button-ripple-uninitialized=""
            mat-icon-button=""
            maticonsuffix=""
          >
            <span
              class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
            />
            <mat-icon
              aria-hidden="true"
              class="mat-icon notranslate material-symbols-outlined add-icon material-icons mat-ligature-font mat-icon-no-color"
              data-mat-icon-type="font"
              role="img"
            >
               add_outline 
            </mat-icon>
            <span
              class="mat-mdc-focus-indicator"
            />
            <span
              class="mat-mdc-button-touch-target"
            />
          </button>
        </div>
        
      </div>
      <div
        class="mdc-line-ripple mdc-line-ripple--deactivating"
        matformfieldlineripple=""
      />
      
    </div>
    <div
      class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
    >
      
      <div
        class="mat-mdc-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
        style="opacity: 1; transform: translateY(0%);"
      >
        
        <div
          class="mat-mdc-form-field-hint-spacer"
        />
      </div>
      
    </div>
  </mat-form-field>
</div>
`;
